_posts/2023-08-01-tmt-cross-project-testing.md

```diff
@@ -0,0 +1,131 @@
+
+As Cockpit is sitting pretty much at the top of the dependency tree, we feel that pain a lot: we find [hundreds of regressions](https://github.com/cockpit-project/bots/issues?q=is%3Aissue+label%3Aknownissue) through our tests. We also [track them in bugzilla](https://bugzilla.redhat.com/buglist.cgi?bug_status=__open__&bug_status=__closed__&f1=status_whiteboard&list_id=12879603&o1=substring&query_format=advanced&v1=CockpitTest) now. Each time that happens, it often takes hours to find the responsible package update, create a standalone reproducer, report a bug, and then either finding a workaround or creating a ["naughty pattern"](https://github.com/cockpit-project/bots/tree/main/naughty) to ignore the problem -- but of course that only takes care of unbreaking our test runs, users are still affected by the regression.
+
+Unfortunately it is technically hard to actually do reverse dependency testing in general in Fedora [†](#footnotes). But also, it is too late at the distro level anyway: At that point the new upstream release which includes the regresssion was already done, and the culprit landed possibly weeks ago already, so that it disappeared from every developer's minds already.
+
+It's [far more efficient](https://en.wikipedia.org/wiki/Shift-left_testing) to flag regressions in a project's consumers right in the pull request that introduces them, *before* it lands: That is the cheapest and most effective time to know what change caused this, the context is in the developer's head, and it avoids the weeks-long cycle of finding the regression, bugzilla, and finding some time to actually fix it. 10 of our known issues have been open since [2021 or earlier](https://github.com/cockpit-project/bots/issues?q=is%3Aissue+is%3Aopen+label%3Aknownissue+created%3A%3C2022-01-01)! [‡](#footnotes)
+
+## Packit and tmt to the rescue
+
+It turns out that modern [Packit](https://packit.dev/) and [tmt](https://tmt.readthedocs.io/) know enough tricks to actually do this. The high-level strategy is:
+
+ 1. Configure Packit to build every commit of your main branch, i.e. every merged PR, into a [COPR](https://copr.fedorainfracloud.org/) repository. That is not just useful for CI, but also for human users: They can easily test the latest fixes or features before you officially release them.
+
+ 2. Identify some strategic dependencies of your project which potentially or actually break your tests often.
+
+ 3. For each of these, add a tmt test plan for your project, and configure Packit to run it for each PR. These run as a separate status, so it makes it very clear if/when a proposed change in the dependency breaks your project.
```

Comment:
```suggestion
 3. For each of these, add a tmt test plan for your project, and configure a separate Packit test job to run it for each PR. You will configure it to run with a dedicated status name, so it makes it very clear if/when a proposed change in the dependency breaks your project.
```

Reply:
This feels like too much technical detail for this high-level strategy. I pointed this out in the "plug it all together" part of the "Enable your test in a dependency project" section instead.

Reply:
agreed, thanks!